### PR TITLE
fix: recover a dropped wj:children close marker on soft nav (#994)

### LINKED ISSUE #994: dogfood: soft-nav to /blog drops the top navbar across browsers (#936 residual)
## Problem

After the #936 fix, CSS no longer drops on a client-router soft navigation, but the **top navbar (a layout DOM element) still randomly fails to survive a soft nav to `/blog`**, specifically on **Android Chrome**. It does not reproduce on desktop. This is a residual of the same #936 investigation, whose `?diag=` harness was originally built for an Android soft-nav bug and even reported navbar state (#939), but the shipped fixes (`7d3a5141` body-context fragment parse, `ffc63bc0` never-strip-stylesheets) only closed the CSS-strip symptom. The layout-element-drop symptom on Android is still open.

The observable failure: navigate to `/blog` via a client-router soft nav on Android Chrome and the persistent top navbar disappears from the DOM (a full reload brings it back). "Random" suggests a parser/timing race, not a deterministic wrong-branch.

## Design / approach

The navbar is part of the outer layout, so this is about **invariant #4: preserve outer-layout DOM identity** (replace only the deepest shared layout's children slot, never the outer chrome). The prime suspects, given the Android-only + #936-adjacent shape:

- **`wj:children` marker detection diverging on Android Chrome's HTML parser.** #936 root cause was that a document-context parse hoisted the leading `wj:children` marker out of `<body>`, defeating the shared-slot match and forcing a **destructive full-body swap** (which is exactly what drops the navbar). The fix moved to `body.setHTMLUnsafe` (body-context parse). If Android Chrome's `setHTMLUnsafe` / Declarative Shadow DOM handling parses the streamed fragment differently (comment-node placement, marker split across text nodes, DSD template timing), the shared-layout match can still miss on Android and fall back to the destructive path, which removes the navbar.
- Confirm whether the Android failure is hitting the **destructive full-body fallback** (`applySwap` L2513) vs the intended **marker swap** (`applySwap` L2496). If it is falling back, the navbar loss is a symptom and the real bug is the marker/shared-path match on Android.
- Cross-check against the known iOS/mobile-WebKit repaint class (#610) but this is Chrome, so more likely a parser/marker issue than a compositor repaint.

Reproduction is the hard part (Android-only, "random"). Reuse the on-device `?diag=` A/B harness pattern from the #936 chore commits (cb49779f, 277437fb, 29f2114c, e60e6fd8) which already reported `wj:children` open/close marker counts + navbar state in a `?diag=` badge; that harness was removed in #946 but is recoverable from git history and is the proven way to diagnose an Android-only soft-nav bug on a real device.

## Implementation notes (for the implementing agent)

- Where to look: `packages/core/src/router-client.js`.
  - `applySwap` (~L2311): the three swap paths. Confirm which one Android takes: frame (L2455), **deepest-shared-layout marker swap (L2496, the correct path)**, or **destructive full-body fallback (L2513, the navbar-dropping path)**.
  - `parseHTML` (~L102, uses `body.setHTMLUnsafe`): the #936 body-context fix. This is the most likely Android-divergent spot.
  - `collectChildrenSlots` (~L724), `longestSharedPath` (~L772), `swapMarkerRange` (~L2576): the shared-layout match that must succeed for the navbar to be preserved.
- Landmines:
  - The `X-Webjs-Have` reduced-head means "absent != removed"; do not confuse a missing head element with a removed one (that was the #936 CSS half via `isPersistentHeadStyle` L3282). This issue is the **body/layout** half, a different code path.
  - "Random" strongly implies a parse/stream-timing race; a single desktop repro attempt passing does NOT clear it. Test on a real Android Chrome device or an Android-Chrome lane, not just Playwright chromium on Linux.
  - The `?diag=` harness was deleted in #946; recover it from the #936 chore commits rather than rebuilding from scratch.
- Invariants to respect: AGENTS.md router invariant "preserve outer-layout DOM identity"; never destructively full-body swap when a shared layout marker exists.
- Tests + docs: add a router browser test under `packages/core/test/routing/browser/` asserting the outer-layout navbar node retains identity across a soft nav (extend the `destructive-fallback` / `partial-fragment-css` #936 suites). Add an e2e case in `test/e2e/e2e.test.mjs` (gated `WEBJS_E2E=1`). If a browser-specific behaviour is discovered, note it in `agent-docs/advanced.md` (client router section) alongside the #610 mobile-WebKit note.

## Acceptance criteria

- [ ] Root cause identified: whether Android Chrome soft-nav to `/blog` falls into the destructive full-body swap and why the `wj:children` shared-layout match misses there
- [ ] The top navbar (outer-layout DOM) retains node identity across a soft nav to `/blog` on Android Chrome
- [ ] A browser/e2e test reproduces the navbar-drop and fails when reverted (counterfactual), covering the outer-layout-identity invariant
- [ ] The test runs in a lane that actually exercises the Android-divergent path (real device or Android-Chrome lane), not desktop-only
- [ ] `agent-docs/advanced.md` updated if a browser-specific constraint is found

### PR DESCRIPTION
**Pull request by vivek7405**
Closes #994

## Summary

A client-router soft nav to `/blog` randomly dropped the persistent top navbar (a full reload brought it back). This is the #936 residual: the CSS-strip half was fixed (never strip stylesheets on a head merge), but the outer-layout drop stayed open.

It surfaced first on Android Chrome and was later reproduced on desktop laptop Chromium (on webjs.dev), so it is a parse/timing race, NOT a device quirk: it is far more frequent under mobile CPU/memory pressure and rare on faster hardware. The fix keys off the symptom, never a user-agent check, so it covers every engine uniformly.

## Root cause

SSR wraps each layout's children in a `<!--wj:children:<path>-->...<!--/wj:children-->` comment pair, and `collectChildrenSlots` needs BOTH comments to register a swappable slot. The `?diag=` harness (#939/#940) showed the live body still carried the OPEN marker (`markers:1`) yet the router still fell to the destructive full-body swap. That only happens when no slot pairs, i.e. the browser's HTML parser intermittently DROPS the trailing `<!--/wj:children-->` close comment. With no pairable slot, `longestSharedPath` returns null and `applySwap`'s path-3 `replaceChildren` wipes the whole body, navbar included.

## Fix

Three commits of logic:

1. **Recover the orphaned open marker.** `collectChildrenSlots` gains an opt-in `recoverOrphans` flag: an OPEN marker left unpaired (its close dropped) is registered with `end: null` ("children run to the parent end"). Wired into the swap-decision (`here`/`there`) and the `X-Webjs-Have` builder. The correct scoped swap then runs and the navbar (which sits BEFORE the open marker) keeps its DOM identity.
2. **Bound the recovered range so trailing chrome survives.** When a layout puts a `<footer>` in the marker's OWN parent, `end: null` would sweep it (live orphan) or duplicate it (incoming orphan). `boundRecoveredEnds` bounds a recovered orphan against the well-formed side's trailing-sibling count. A zero count (the shipped `<main>${children}</main>` idiom) keeps the sweep-to-parent-end unchanged.
3. **Route orphaned navs through the full-page path.** `buildHaveHeader` reports STRICTLY paired markers, so an orphaned page omits the layout from `have` and the server returns the full page (trailing chrome present) rather than a reduced fragment where the bounding could not align. Plus a degenerate `cut<=0` fallback.

## Scope

Verified webjs.dev is in the fully-covered shape: a single `/` layout level, the `.site-top` navbar before the `wj:children:/` open marker, so a dropped close on either side falls into the fixed path. The one uncovered shape (a dropped INNER close in a NESTED layout) does not apply to webjs.dev and is filed as follow-up #1003; the wrap-`${children}` idiom is the mitigation.

## Tests

- **Unit** (`packages/core/test/routing/router-client.test.js`): recovery (orphaned open registers `end:null`; strict-default counterfactual precondition; nested dropped-outer-close), the `_applySwap` navbar-identity case, both orphan directions with a trailing footer (no sweep, no duplication), a strict-`have` assertion, and the degenerate `cut===0` boundary.
- **Browser** (`packages/core/test/routing/browser/orphaned-marker-navbar.test.js`, Chromium/Firefox/WebKit): navbar identity across a dropped-close soft swap, a well-formed swap, `buildHaveHeader` strict pairing, and trailing-footer preservation. Reverting the recovery fails the navbar case; reverting the bounding fails the footer case; reverting the strict-`have` fails its case (clean counterfactuals on all three engines).
- **E2E** (`test/e2e/e2e.test.mjs`): navbar-identity across a real soft nav, passing in CI against freshly built dist.

## Docs

- `.agents/skills/webjs/references/client-router-and-streaming.md` and `docs/app/docs/client-router/page.ts`: the dropped-close recovery, framed as a browser-agnostic parse race, plus the trailing-content bounding and the wrap-`${children}` safe idiom.

## Test plan
- [x] unit (core routing) green (168)
- [x] browser (core routing) green on Chromium/Firefox/WebKit, with counterfactuals
- [x] e2e navbar-identity case passes in CI
- [x] dogfood: all-four-apps CI job green; website / docs / ui-website boot 200 locally

### CODE REVIEW

**Comment by vivek7405** [comment]
**Design rationale: why recover the orphaned open marker, not touch path 3**

Walking the #936 history first, because it shapes this fix. #936 had two halves. The CSS-strip half was closed by making `mergeHead` never remove a stylesheet on a soft nav (the persistent-CSS model). The body/layout half (the navbar drop) was NOT closed: an earlier attempt to make `applySwap`'s path-3 fallback do a real full-page load was reverted in the same PR (#945), because path 3 is a legitimate normal soft-nav path (the blog uses it for cross-layout navs), so reloading it regressed client nav and mis-fired inside popstate. So path 3 is off-limits for a "reload on no-shared-marker" fix, and any defensive change there has to distinguish a genuine root-layout change (where wiping the outer chrome is CORRECT) from a dropped-close-comment false negative (where it is the bug). That distinction is not reliably knowable at the path-3 decision point.

So I fixed it one level up, at marker discovery. The diag harness (#939/#940) established the live body keeps its OPEN `wj:children` marker yet the router still falls to path 3, which only happens when `collectChildrenSlots` cannot PAIR a slot. The only way that happens with the open present is a dropped CLOSE comment. Recovering the orphaned open (registering it with `end: null`, "children run to the parent end") makes the shared-path match succeed, so the correct SCOPED swap (path 2) runs and the navbar, which sits before the open marker, is never in the swap range. This needs no knowledge of why the device dropped the comment, so it is browser-agnostic and deterministic to test.

Two things I checked make `end: null` safe rather than a hack: `swapMarkerRange`'s slice loops already terminate at the natural sibling end when `end` is null, and `reconcileSiblings` inserts before `null`, which appends at the parent end. So a recovered range swaps the whole children region with no close marker, and a trailing already-executed boot `<script>` swept into the range is harmless (removing an executed module element does not un-run it). I deliberately left the flag OFF by default on the exported `collectChildrenSlots` so its strict pairing contract is unchanged; only the swap-decision (`here`/`there`) and the `X-Webjs-Have` builder opt in (the latter so a dropped close does not make the client under-report a layout it has and pull a full page instead of the reduced fragment).

One observation from testing worth recording: with the recovery reverted, the node/linkedom `_applySwap` navbar test HANGS in path 3 rather than failing cleanly (a linkedom-only quirk in the full-body `mergeHead`/`replaceChildren`/MutationObserver interaction on a content-ful body). Real browsers do not hang: the same counterfactual in Chromium/Firefox/WebKit fails cleanly (navbar node replaced). So the deterministic counterfactual lives in the browser test and in the collect-level "orphaned open is NOT paired by default" unit assertion, not in the applySwap node test.

**Review by vivek7405 — COMMENTED** [review]
First review pass, focused on the correctness of the orphan recovery. One real content-loss bug in two directions, plus the test gap that let it through.

The root layout emits the footer (and the boot script) as siblings AFTER the `<!--/wj:children-->` close marker, so `end: null` ("run to the parent end") over-reaches whenever the marker's own parent holds trailing chrome. The blog and scaffold happen to wrap `${children}` in `<main>` (footer outside), which is why the headline case is fine, but an unwrapped layout would lose or duplicate its footer. Fixed by bounding a recovered orphan against the well-formed side's trailing-sibling count, with unit + browser coverage for both directions.

File: packages/core/src/router-client.js [hunk 1]
```diff
@@ -2493,19 +2522,33 @@ function applySwap(doc, frameId, revalidating, href, incomingBuild, incomingSrc)
     return;
   }
 
-  // 2. Auto-derived layout-marker swap.
-  const here = collectChildrenSlots(document.body);
-  const there = collectChildrenSlots(doc.body);
+  // 2. Auto-derived layout-marker swap. Recover an orphaned open marker on
+  // either side (#994): a dropped close comment must not force the destructive
+  // path-3 fallback that wipes the outer layout (navbar).
+  const here = collectChildrenSlots(document.body, { recoverOrphans: true });
```
Comment (vivek7405) [inline packages/core/src/router-client.js:2542]: Recovering on the LIVE side with `end: null` sweeps to the marker's parent end. When a `<footer>` sits after the (dropped) close in that same parent, the live footer lands in the swap range and the incoming (well-formed, bounded) side does not replace it, so it is removed and never restored. Symmetric case on the incoming side duplicates the footer into the children region.

File: packages/core/test/routing/router-client.test.js [hunk 2]
```diff
@@ -3321,6 +3380,136 @@ test('applySwap does NOT evict when the app-source id is unchanged (no churn)',
   }
 });
 
+test('applySwap: a dropped incoming close marker still scoped-swaps and keeps the navbar node (#994)', () => {
```
Comment (vivek7405) [inline packages/core/test/routing/router-client.test.js:3383]: Every recovery test body is `nav + open + main + [close]` with nothing after the close marker, so the destructive interaction with trailing outer-layout content (footer/scripts) in the marker's parent is unexercised, and the live-side-orphan direction never runs through swapMarkerRange.

File: packages/core/src/router-client.js (hunk 1 above)
Comment (vivek7405) [reply packages/core/src/router-client.js:2542]: Fixed in [`110377c7`](https://github.com/webjsdev/webjs/commit/110377c7). Added unit tests for both orphan directions with a trailing `<footer>` in the marker's parent (asserting no sweep and no duplication), a real-browser footer-preservation case, and an e2e navbar-identity case.

File: packages/core/test/routing/router-client.test.js (hunk 2 above)
Comment (vivek7405) [reply packages/core/test/routing/router-client.test.js:3383]: Fixed in [`110377c7`](https://github.com/webjsdev/webjs/commit/110377c7). Added unit tests for both orphan directions with a trailing `<footer>` in the marker's parent (asserting no sweep and no duplication), a real-browser footer-preservation case, and an e2e navbar-identity case.

**Comment by vivek7405** [comment]
**Verification: how the fix is proven, and the worktree e2e caveat**

The client behaviour is proven where it matters, in a real browser. The core routing browser suite (Chromium, Firefox, WebKit) covers both halves with clean counterfactuals: reverting the orphan recovery makes the navbar-identity case fail (navbar node replaced), and reverting the trailing-content bounding makes the footer-preservation case fail (footer swept), both on all three engines while the well-formed cases stay green. Unit coverage adds both orphan directions with a trailing footer and the collect-level "orphaned open is not paired by default" precondition.

The e2e suite has a worktree-only caveat worth recording so nobody mistakes it for a regression. This branch is a git worktree, so its blog serves the prebuilt `dist/webjs-core-browser.js` from the shared install, which predates this change and every recent client-router fix. So the client-nav e2e cases (navbar, header, counter, permanent-identity) fail against that stale bundle. I confirmed it is the bundle, not the change, two ways: (1) a baseline e2e run with this branch's source reverted produced the IDENTICAL 7-failure set, so the change adds zero regressions; (2) serving the worktree source per-file instead broke the blog wholesale (the known cjs-lexer per-package-symlink trap), so src-mode is not a usable local lane here. The added e2e navbar-identity case is correct by construction (it mirrors the existing #250 permanent-identity test) and runs against a freshly built dist in CI, where the Build job compiles this branch's source into the bundle.

Dogfood: website / docs / ui-website boot 200 in dist mode with no broken modulepreloads. The blog is the e2e target above.

**Review by vivek7405 — COMMENTED** [review]
Second pass, on the bounding fix itself. Two issues, one of them the interesting kind where the fix undercut itself.

The important one: recovering the orphan in `buildHaveHeader` made an orphaned page send `have=/`, so the server returned a reduced marker-pair-only fragment with no trailing chrome, which is exactly the `tail === 0` case the bounding cannot cover, and it was the DOMINANT path for an orphaned nav. Reporting strict pairs instead makes an orphaned page omit the layout, so the server returns the full page and the bounding aligns against real trailing content. The navbar preservation never depended on the `have` recovery (it comes from applySwap), so this is a clean win for the cost of a full page on a rare nav. Also fixed the degenerate `cut <= 0` branch, which returned an empty range (blank/duplicate) instead of a full-region swap.

File: packages/core/src/router-client.js [hunk 3]
```diff
@@ -1047,6 +1073,15 @@ async function performSubmission(href, method, body, frameId, form) {
...
+  // Strict pairing here, deliberately (#994): if the live page's close marker was
+  // dropped, its orphaned open is NOT reported, so `have` omits that layout and
+  // the server returns the FULL page (with the outer layout's trailing chrome)
+  // rather than a reduced marker-pair-only fragment. `applySwap` then recovers
+  // the orphan and bounds the swap against that full page's trailing-sibling
+  // count, preserving BOTH the navbar and the footer for any layout. Recovering
+  // here would send `have=/`, get back a reduced fragment with no trailing chrome
+  // (`tail === 0`), and sweep an unwrapped layout's footer. The extra bytes of a
+  // full page on the rare orphaned nav are the price of that correctness.
   const slots = collectChildrenSlots(document.body);
```
Comment (vivek7405) [inline packages/core/src/router-client.js:1093]: Recovering the orphan here sent `have=/`, so the server returned a reduced fragment with no trailing chrome (`tail === 0`), and the swap swept an unwrapped layout's footer. That reduced-fragment path was the dominant path for an orphaned nav, so the bounding never got real trailing content to align against.

File: packages/core/src/router-client.js [hunk 4]
```diff
@@ -2562,15 +2611,89 @@ function blurOutgoingFocus() {
...
+  const nodes = [];
+  for (let n = orphanStart.nextSibling; n; n = n.nextSibling) nodes.push(n);
+  // Preserve the last `tail` nodes by stopping before index `cut`. A `cut` at or
+  // below zero is a degenerate mismatch (the well-formed side claims more
+  // trailing siblings than the orphan has nodes at all, which a same-layout nav
+  // should never produce): fall back to `null` (sweep to the parent end, a normal
+  // full-region swap) rather than an empty range, which would duplicate or blank
+  // the children.
+  const cut = nodes.length - tail;
+  if (cut <= 0) return null;
```
Comment (vivek7405) [inline packages/core/src/router-client.js:2690]: The degenerate branch (well-formed side claims more trailing siblings than the orphan has nodes) returned `orphanStart.nextSibling` as an EXCLUSIVE end, an empty range: it duplicates the children (live orphan) or blanks them (incoming orphan), the opposite of the 'keep everything' the comment claimed.

File: packages/core/src/router-client.js (hunk 3 above)
Comment (vivek7405) [reply packages/core/src/router-client.js:1093]: Fixed in [`edb5e84a`](https://github.com/webjsdev/webjs/commit/edb5e84a). The `cut <= 0` branch now returns `null` (sweep to parent end, a normal full-region swap) instead of an empty range. Added a degenerate-mismatch applySwap test.

File: packages/core/src/router-client.js (hunk 4 above)
Comment (vivek7405) [reply packages/core/src/router-client.js:2690]: Fixed in [`edb5e84a`](https://github.com/webjsdev/webjs/commit/edb5e84a). The `cut <= 0` branch now returns `null` (sweep to parent end, a normal full-region swap) instead of an empty range. Added a degenerate-mismatch applySwap test.

File: packages/core/test/routing/router-client.test.js [hunk 5]
```diff
@@ -3321,6 +3380,181 @@ test('applySwap does NOT evict when the app-source id is unchanged (no churn)',
...
+    assert.ok(globalThis.document.getElementById('new5'), 'the children slot swapped');
+    assert.ok(!globalThis.document.getElementById('old5'), 'old children replaced');
+  } finally {
+    globalThis.location = savedLocation;
+    globalThis.document.head.innerHTML = savedHead;
+    globalThis.document.body.innerHTML = savedBody;
+  }
+});
+
+test('applySwap: a degenerate trailing-count mismatch sweeps to parent end, never blanks or duplicates (#994)', () => {
```
Comment (vivek7405) [inline packages/core/test/routing/router-client.test.js:3513]: This exercised cut<0 (tail>node-count), where nodes[cut] is undefined and behaves like the null the guard returns, so the boundary case cut===0 (where the empty-range bug actually bites) went uncovered.

File: packages/core/src/router-client.js [hunk 6]
```diff
@@ -1047,6 +1073,18 @@ async function performSubmission(href, method, body, frameId, form) {
...
+  // chrome) rather than a reduced marker-pair-only fragment. `applySwap` then
+  // recovers the orphan and bounds the swap against that full page's
+  // trailing-sibling count, preserving the navbar and the footer. Recovering here
+  // would send `have=/`, get back a reduced fragment with no trailing chrome
+  // (`tail === 0`), and sweep an unwrapped layout's footer. The extra bytes of a
+  // full page on the rare orphaned nav are the price of that correctness. A
+  // dropped INNER close in a nested layout is a separate, unfixed mispairing
+  // limitation (the surviving outer close pairs with the inner open); the wrapped
+  // `${children}` idiom keeps it harmless.
   const slots = collectChildrenSlots(document.body);
```
Comment (vivek7405) [inline packages/core/src/router-client.js:1093]: The comment says the full-page fallback preserves navbar+footer 'for any layout', but a dropped INNER close in a nested layout mispairs (outer close pairs with inner open) and still sweeps outer trailing chrome. The claim needs scoping to the sole/outermost dropped close.

**Review by vivek7405 — COMMENTED** [review]
Third pass, on the round-2 fixes. Two things: a test-clarity nit and an overclaim in the new comments/docs. Both handled in ae9726be. Also filed the pre-existing nested-layout mispairing this surfaced as #1003.

The degenerate-mismatch test did catch the real revert (I verified it fails when the guard goes back to `return orphanStart.nextSibling`), but it sat at cut<0 where the boundary behaviour is muddy, so I moved it to the exact cut===0 boundary where `nodes[cut]` IS `orphanStart.nextSibling`, so any non-null return produces the empty range and the test fails under either revert model. The bigger one: the new comments and the docs note claimed the full-page fallback preserves navbar+footer 'for any layout', which is false for a dropped INNER close in a nested layout (the surviving outer close mispairs with the inner open, both ends non-null, so bounding is skipped). Scoped the wording to the sole/outermost case and filed the nested mispairing as #1003 (it is pre-existing pop-logic, not something this PR introduces or worsens).

File: packages/core/test/routing/router-client.test.js (hunk 5 above)
Comment (vivek7405) [reply packages/core/test/routing/router-client.test.js:3513]: Fixed in [`ae9726be`](https://github.com/webjsdev/webjs/commit/ae9726be). Scoped the buildHaveHeader/boundRecoveredEnds comments and the docs note to the sole/outermost dropped close. The nested inner-close mispairing is pre-existing pop-logic (not introduced here) and is filed as #1003 with a path-on-close-marker fix sketch.

File: packages/core/src/router-client.js (hunk 6 above)
Comment (vivek7405) [reply packages/core/src/router-client.js:1093]: Fixed in [`ae9726be`](https://github.com/webjsdev/webjs/commit/ae9726be). Scoped the buildHaveHeader/boundRecoveredEnds comments and the docs note to the sole/outermost dropped close. The nested inner-close mispairing is pre-existing pop-logic (not introduced here) and is filed as #1003 with a path-on-close-marker fix sketch.

File: packages/core/src/router-client.js [hunk 7]
```diff
@@ -2299,7 +2337,7 @@ function findInSlice(slice, id) {
  * swap settles. The View Transitions API snapshots and replaces DOM, so
  * elements can need a re-upgrade once the animation finishes.
  *
- * @param {{ start: Comment, end: Comment } | undefined} range
+ * @param {{ start: Comment, end: Comment | null } | undefined} range
  */
```
Comment (vivek7405) [inline packages/core/src/router-client.js:2346]: This still declares end as Comment, but it now receives the same range object that can carry end: null (a recovered tail=0 or degenerate orphan). The sibling swapMarkerRange/reconcileSiblings JSDocs were updated to Comment | null; this one was missed.

**Review by vivek7405 — COMMENTED** [review]
Convergence check. The recovery and bounding logic is correct and internally consistent across all four directions (live-orphan, incoming-orphan, both-orphan, well-formed), buildHaveHeader's strict pairing matches the bounding rationale, the null-end is handled by swapMarkerRange/reconcileSiblings, and the cut===0 test has real counterfactual value. One cosmetic annotation gap remained, fixed in [`c4ba3ad3`](https://github.com/webjsdev/webjs/commit/c4ba3ad3). No invariant-11 glyph issues, no new .ts in packages, JSDoc-only preserved.

File: packages/core/src/router-client.js (hunk 7 above)
Comment (vivek7405) [reply packages/core/src/router-client.js:2346]: Fixed in [`c4ba3ad3`](https://github.com/webjsdev/webjs/commit/c4ba3ad3).

**Review by vivek7405 — COMMENTED** [review]
Final pass, clean. Null-end handling is sound end to end (the slice loops and reconcileSiblings append at parent end), the cut===0 fallback matches its counterfactual test, strict buildHaveHeader is consistent with recover-at-apply, and the docs no longer overclaim (the nested inner-close case is scoped out and filed as #1003). Coverage is complete across unit, browser, and e2e for every behaviour the change introduces. Nothing left to fix.

**Comment by vivek7405** [comment]
**Scope confirmation: the fix is browser-agnostic, verified against live webjs.dev**

The navbar drop reproduced on desktop laptop Chromium on webjs.dev, having previously only been seen on Android Chrome. That does not narrow or move the fix, it confirms the framing: the close-comment loss is a parse/timing race, not a device quirk, so the recovery is the right shape because it keys off the SYMPTOM (a missing close forcing the path-3 full-body swap), never a user-agent check.

I checked the live site to be sure the desktop case is the same root cause and inside the covered shape:
- `.site-top` (navbar) is at byte 14953, before the `<!--wj:children:/-->` open marker at 21275, so the navbar is outside the scoped-swap range and only the full-body fallback can wipe it.
- Live `/blog` emits a SINGLE marker level (`<!--wj:children:/-->`); there is no nested layout under `/blog`. So a dropped close here is the sole-close case the fix covers fully, and the nested inner-close gap (#1003) does not apply.
- On a same-layout soft nav, path-3 (navbar loss) is only reachable when marker pairing fails on the live or incoming side, both of which the recovery + full-page routing handle.

Why it looked Android-only: the race is far more frequent under mobile CPU/memory pressure and rare on faster hardware, so desktop hitting it is the rare tail of the same race. Retitled the issue and reframed the comments/docs to drop the Android-specific language.

File: packages/core/src/router-client.js [hunk 8]
```diff
@@ -2562,15 +2619,92 @@ function blurOutgoingFocus() {
...
+ * Compute the bounded end sentinel (exclusive) for an orphaned open marker,
+ * preserving `tail` trailing siblings at the end of the marker's parent, where
+ * `tail` is the well-formed counterpart's trailing-sibling count (#994).
+ *
+ * @param {Comment} orphanStart  The orphan's open marker.
+ * @param {Comment} wellFormedEnd  The counterpart's real close marker.
+ * @returns {Comment | Node | null}  A node to stop before, or null (sweep to end).
+ */
+function boundOrphanEnd(orphanStart, wellFormedEnd) {
+  const tail = trailingSiblingCount(wellFormedEnd);
```
Comment (vivek7405) [inline packages/core/src/router-client.js:2677]: Low severity: `trailingSiblingCount` counts ALL node types (text/comment/element), so the `tail` measured on the well-formed side and the orphan's own trailing-node count must match exactly for the cut to land right. For an UNWRAPPED layout with an asymmetric orphan, a single trailing whitespace-text-node difference between the live and incoming parses (e.g. a footer component that mutated structure post-SSR) would shift the cut by one and sweep the footer's leading node or leave a stale child. Both sides come from the same SSR string so counts normally match, and the wrapped idiom makes tail=0 (no cut), so this is a narrow risk. An element-aware count would harden it if you want to close it fully.

**Review by vivek7405 — COMMENTED** [review]
Full-diff review at high effort. After five prior rounds the logic holds up: recovery is applied only in applySwap (where null-end is handled), the strict callers stay strict on purpose (renderInPlaceNavError at router-client.js:1626 derefs deepest.end.parentNode, so recovering there would risk a null deref, and applyOptimisticLoading degrades to no skeleton, both safe), slot objects are fresh per nav, and every null-end consumer (swapMarkerRange loops, reconcileSiblings insert-before-null, the upgrade loop) is correct. One low-severity in-diff observation inline below, mitigated by the shipped wrap-`${children}` idiom. One cosmetic note path-level: renderInPlaceNavError's local `deepest` is still annotated `end: Comment` while collectChildrenSlots now returns `end: Comment | null` (no runtime impact, strict caller; the annotation correctly documents the strict guarantee). Neither blocks merge.

File: packages/core/src/router-client.js (hunk 8 above)
Comment (vivek7405) [reply packages/core/src/router-client.js:2677]: Leaving as-is for this PR (rejected, not blocking): this only affects FOOTER preservation in an UNWRAPPED layout, never the navbar (which sits before the open marker and is preserved by the scoped swap the moment the orphan is recovered, independent of any tail counting). The shipped wrap-`${children}`-in-`<main>` idiom makes tail=0 so real apps never hit the cut at all, and both sides parse from the same SSR string so counts match unless a footer component mutates structure post-hydration. Hardening it (element-aware counting) is worth a follow-up if an unwrapped-layout app ever needs it, but adding it here after the review converged would be risk for no real-app benefit.

File: test/e2e/e2e.test.mjs [hunk 9]
```diff
@@ -506,9 +506,9 @@ describe('E2E: Blog example', { skip: !process.env.WEBJS_E2E && 'set WEBJS_E2E=1
   test('a streaming-RPC action renders tokens incrementally (#489)', async () => {
     // <token-stream> calls an async-generator action; each yielded token streams
     // over the single RPC response and is appended as it arrives. We assert (1)
-    // the action response is the framed stream content type, and (2) the rendered
-    // count climbs over time (a snapshot mid-stream is below the final count),
-    // proving incremental arrival rather than one buffered result.
+    // the action response is the framed stream content type, and (2) the FIRST
```
Comment (vivek7405) [inline test/e2e/e2e.test.mjs:509]: The header comment described the old 'snapshot mid-stream is below the final count' approach, which the de-flake removed. Updated to describe the first-paint MutationObserver check.

**Review by vivek7405 — COMMENTED** [review]
Reviewed the de-flake (the streaming-RPC first-paint MutationObserver). The approach is race-free and keeps the counterfactual: light-DOM `.ts-item`s under the observed element so subtree catches them, the observer is installed before the click, the n>0 guard skips the empty initial render, a buffered flush first-paints all 8 so first===final fails, and the 60ms server spacing guarantees the first paint shows <8 on a healthy stream. Observer is disconnected and the next test's page.goto discards the globals, so no cross-test leak. One stale header comment still described the removed mid-snapshot approach, fixed in [`415b6b94f47a6f2d5dd27d30f5d67207ecff3235`](https://github.com/webjsdev/webjs/commit/415b6b94f47a6f2d5dd27d30f5d67207ecff3235).

File: test/e2e/e2e.test.mjs (hunk 9 above)
Comment (vivek7405) [reply test/e2e/e2e.test.mjs:509]: Fixed by updating the comment to describe the first-paint check.

**Review by vivek7405 — COMMENTED** [review]
Final pass, clean. The orphan-recovery + bounding logic, the strict-have rationale, the null-end handling, all the unit/browser tests with their counterfactuals, and the race-free streaming e2e rewrite all hold up, and the comments/docs match the code. Nothing left.
